feat(data-catalog): Adding Hue

tumido opened the pull request:
Adding Hue component Data catalog.

Part of: https://github.com/opendatahub-io/odh-manifests/issues/222, https://github.com/opendatahub-io/odh-manifests/issues/105, [DATAHUB-2294](https://projects.engineering.redhat.com/browse/DATAHUB-2294)

Based on reference implementation in https://github.com/AICoE/idh-manifests/pull/27 for Internal DH.

It will need further cleanup and extraction of parts into overlays later on (storage class, externalize the database). This is expected to happen in consecutive PRs.

tumido commented:
cc @maulikjs, @rimolive

tumido commented:
/retest

tumido commented:
/retest

test failures not related to the code change

anishasthana commented:
/retest

tumido commented:
/retest

rimolive commented on hue/hue/base/hive-site-xml-secret.yaml:
```diff
@@ -0,0 +1,38 @@
+
+      <property>
+        <name>fs.s3a.awsAccessKeyId</name>
+        <value></value>
```
Shouldn't this line use a variable to populate the S3 Access Key

tumido replied:
@rimolive all requested changes should be fixed now, please check again :+1:

rimolive commented on hue/hue/base/hive-site-xml-secret.yaml:
```diff
@@ -0,0 +1,38 @@
+
+      <property>
+        <name>fs.s3a.awsSecretAccessKey</name>
+        <value></value>
```
Same as above

rimolive commented on hue/hue/base/hue-ini-secret.yaml:
```diff
@@ -0,0 +1,179 @@
+        port=3306
+        user=datacatalog
+        password=datacatalog
+        name=datacatalog
```
These lines should be parametrized or at least generate a random password.

rimolive commented on hue/hue/base/hue-ini-secret.yaml:
```diff
@@ -0,0 +1,179 @@
+      [[aws_accounts]]
+        [[[default]]]
+          access_key_id=
+          secret_access_key=
```
These lines should be populated

rimolive commented on hue/hue/base/params.env:
```diff
@@ -0,0 +1,4 @@
+storage_class=
+s3_endpoint=
+s3_region=
+s3_is_secure=false
```
It is missing the parameters for the S3 keys

tumido replied:
let's discuss this at https://github.com/opendatahub-io/odh-manifests/pull/228#discussion_r521379415 it's the same thing.

tumido replied:
resolved via

```
s3_credentials_secret=thriftserver-sample-s3
```

rimolive commented on hue/hue/base/hue-ini-secret.yaml:
```diff
@@ -0,0 +1,179 @@
+      [[aws_accounts]]
+        [[[default]]]
+          access_key_id=$(s3_key_id)
+          secret_access_key=$(s3_access_key)
```
It is better if you name the variables as `s3_access_key_id` and `s3_secret_access_key_id` as they better represent what we are storing in these variables

tumido replied:
done

vpavlin commented on hue/hue/base/hue-dc.yaml:
```diff
@@ -0,0 +1,70 @@
+            failureThreshold: 4
+            httpGet:
+              path: /
+              port: 8000
```
Should the port 8000 be exposed for the container?

tumido replied:
Fixed

vpavlin commented on hue/hue/base/hue-dc.yaml:
```diff
@@ -0,0 +1,70 @@
+---
+apiVersion: apps.openshift.io/v1
+kind: DeploymentConfig
```
Should we use `Deployment` instead of DC since (IIUC) Deployment is now preferred by OpenShift?

tumido replied:
It would definitely simplify things, I'll look into that. The DC was chosen to mirror Internal DH's and the legacy ODH setup. Would you rather see it changed now and here or in a consecutive PR?

vpavlin replied:
I personally am fine with having the change in this PR, up to you

tumido replied:
Changed to `Deployment`

vpavlin commented on hue/hue/base/hue-ini-secret.yaml:
```diff
@@ -0,0 +1,179 @@
+    [desktop]
+      # Hide unused apps
+      app_blacklist=impala,security,jobbrowser,jobsub,pig,hbase,sqoop,zookeeper,spark,oozie,search
+      secret_key=thisISaSECRET_1234
```
What is this `secret_key`?

tumido replied:
It's a django app secret key

See: https://docs.cloudera.com/documentation/enterprise/5-7-x/topics/cdh_ig_hue_config.html#topic_15_6_2__title_371

Similar to the superser's secret key https://github.com/opendatahub-io/odh-manifests/blob/58b94891d0808fd333271a0a282b018af4e2ab81/superset/base/configmap.yaml#L24

I think we should rather have it loaded from the `params.env`, correct?

tumido replied:
Templated via `params.env`

vpavlin commented on hue/hue/base/hue-mysql-dc.yaml:
```diff
@@ -0,0 +1,114 @@
+---
+apiVersion: apps.openshift.io/v1
+kind: DeploymentConfig
```
Same question - should we use `Deployment`?

tumido replied:
Same answer. :smile: you tell me. :slightly_smiling_face: 

I have no preference. `Deployment` makes it more close to upstream kubernetes and lowers complexity of this resource manifest.

tumido replied:
Changed to `Deployment`

vpavlin commented on hue/hue/base/hue-mysql-dc.yaml:
```diff
@@ -0,0 +1,114 @@
+metadata:
+  name: hue-mysql
+  annotations:
+    template.alpha.openshift.io/wait-for-ready: "true"
```
What is this doing?

tumido replied:
https://docs.openshift.com/container-platform/4.2/openshift_images/using-templates.html#templates-waiting-for-readiness_using-templates

I have no idea if it actually has any meaningful impact for ODH operator though. It comes from the initial implementation in the legacy ODH.

vpavlin commented on hue/hue/base/hue-mysql-dc.yaml:
```diff
@@ -0,0 +1,114 @@
+        containerNames:
+          - mysql
+        from:
+          kind: ImageStreamTag
```
Is this an IS coming from `openshift` namespace? Do we need more control over version?

tumido replied:
If we change this to the `Deployment` this whole section will go away.

~Also it's already controlled in the kustomization (including the namespace).~ You've found that. :)

tumido replied:
Changed to `Deployment`

vpavlin commented on hue/hue/base/hue-mysql-dc.yaml:
```diff
@@ -0,0 +1,114 @@
+                  name: "hue-mysql-secret"
+                  key: database-root-password
+          resources:
+            limits:
```
I think we should set both limits and requests, otherwise (IIRC) if only limits is set, requests==limits

tumido replied:
@rimolive do you have any suggestion on what would be reasonable values here?

tumido replied:
I've just put some random values in there that seemed believable and sensible to me.

vpavlin commented on hue/hue/base/hue-mysql-pvc.yaml:
```diff
@@ -0,0 +1,13 @@
+metadata:
+  name: hue-mysql
+  annotations:
+    volume.beta.kubernetes.io/storage-class: "$(storage_class)"
```
I am not sure if we need this. We had it for JupyterHub because we had one deployment (MOC) which did not set default storage class and then one (not sure, maybe early IDH?) where default was not what we wanted to use.

The problem with having storage class here like this is, that you will always need to provide it - since storage class can have 3 values - `null`, ``""`, actual name...so in this case if `storage_class` is not set, it would end up being `""` and fail deployment in most cases.

You can take a look at JupyterHub - we have an overlay `storageClass` which adds the definitions and parameter if enabled.

WDYT?

tumido replied:
We can put it into an overlay. I actually initially wanted to do that, however I kinda tried to keep the initial PR as compact as possible and not introduce unnecessary complexity. I know the storage class in data catalog is used in the Internal DH with different values for stage/prod, so I think this parameter still applies.

tumido replied:
Moved to overlay

vpavlin commented on hue/hue/base/hue-mysql-secret.yaml:
```diff
@@ -0,0 +1,16 @@
+metadata:
+  name: hue-mysql-secret
+  annotations:
+    template.openshift.io/expose-username: "{.data['database-user']}"
```
How are these used?

tumido replied:
No idea. This comes from the original implementation in the legacy ODH operator.

tumido replied:
removed

vpavlin commented on hue/hue/base/kustomization.yaml:
```diff
@@ -0,0 +1,133 @@
+commonLabels:
+  opendatahub.io/component: "true"
+  component.opendatahub.io/name: hue
+  component.opendatahub.io/part-of: datacatalog
```
I like this:)

tumido replied:
It kinda mimicks the https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/ scheme :)

rimolive replied:
Is this something we need to change? If not, can we mark as Resolved?

tumido replied:
I think it can stay. :wink:

vpavlin commented on hue/hue/base/kustomization.yaml:
```diff
@@ -0,0 +1,133 @@
+    newName: prom/mysqld-exporter
+
+patchesJson6902:
+  - patch: |-
```
Ah, you can partially disregard my comment about the image above:)

But, the problem with using images from `openshift` namespace is that versions disappear between OCP version releases - we hit it every now and then with JupyterHub DB. I would rather us to have control over when we update, then disappearing image forcing us to - can we rather pull the image from RH container catalog?

tumido replied:
:) Sure, we can change the base image.

tumido replied:
removed since `Deployment`s don't support triggers anyways (changed from `DeploymentConfig`

tumido commented:
CI is broken, I'll retest it later.

tumido commented:
/retest

anishasthana commented:
/retest

rimolive commented on hue/hue/base/hive-site-xml-secret.yaml:
```diff
@@ -0,0 +1,77 @@
+
+      <property>
+        <name>fs.s3a.awsSecretAccessKey</name>
+        <value>AWS secret key</value>
```
It's missing the parameter to specify the AWS Secret key from params.env

tumido replied:
This PR is WIP for a reason. Please don't review now.

tumido replied:
Btw. The aim is to load them from environment, see above. It can't be specified here as inlined values, since that would require us traversing vars multiple times and kustomize can't do that. (a parameter defines which secret holds the s3 credentials -> loaded into vars -> based on this var we load another secret and store the creds into another variables and then template this file, that's simply not possible).

It should be doable to load it from the environment, so don't mind these values at all. I should be able to bypass it.

rimolive commented on hue/hue/base/hive-site-xml-secret.yaml:
```diff
@@ -0,0 +1,77 @@
+
+      <property>
+        <name>fs.s3a.endpoint</name>
+        <value>AWS S3 endpoint</value>
```
Same here.

tumido replied:
Irrelevant, closing.

tumido commented:
This is also finally ready. All issues when communicating with Spark are resolved now.

Key updates:
- Huge Hue image upgrade - using 4.8.0 now from my own build, will try to get it into [ODH's quay repo for Hue](https://quay.io/repository/opendatahub/hue). Using multistage build so the image size was effectively reduced from 2.8 GiB (total image size) to 1.3 GiB
- Hue can source it's S3 credentials from environment via custom scripts and additional Hive settings - Similarly to Thrift Server, user can specify the s3 secret which holds their S3 credentials and Hue will source that.
- Simplified container's (Hue runtime and DB init job) commands
- Previously used `livelinessProbe` caused the container to be killed and restarted often because it tried to load the whole UI without any JS cache, switched over to a lightweight `is_alive` endpoint
- Can list S3 files in file browser as well as in S3 application
- Can talk to Thift Server and populate Hive meta store with external/internal tables etc.

Standalone ODH Datacatalog can be deployed via: https://gist.github.com/tumido/6b91a18954041d1f4ba13bc5eeb09d8f

Here are some screenshots:
![image](https://user-images.githubusercontent.com/7453394/100073537-e97b8080-2e3d-11eb-9fe7-6fde299cee4f.png)
![image](https://user-images.githubusercontent.com/7453394/100074851-8985d980-2e3f-11eb-956e-a82ac5c1526b.png)
![image](https://user-images.githubusercontent.com/7453394/100074965-a4584e00-2e3f-11eb-9e4b-b2367bd10ff4.png)
![image](https://user-images.githubusercontent.com/7453394/100076074-14b39f00-2e41-11eb-9e17-436be26242ca.png)
![image](https://user-images.githubusercontent.com/7453394/100077054-2e091b00-2e42-11eb-8480-c529398fa318.png)

tumido commented:
bump @accorvin @rimolive can we have this one reviewed once more?

nakfour commented:
@LaVLaS and @vpavlin please re-review

vpavlin commented:
@tumido I have just noticed the directory structure is `hue/hue/base` - any particular reason why the `hue` dir is doubled?

vpavlin commented on hue/hue/base/hue-db-init-job.yaml:
```diff
@@ -0,0 +1,28 @@
+---
+apiVersion: batch/v1
```
Thinking about this a bit more - what happens when we upgrade and the Jub is reconciled? How do we make it run again? (IIRC Job only runs once and there is no way to retrigger it other than delete and apply again)

tumido replied:
Gotcha. Good point. I think we should rather have it as an init container for Hue then. WDYT? :+1:

tumido replied:
Moved :+1:

vpavlin commented on hue/hue/base/hue-mysql-pvc.yaml:
```diff
@@ -0,0 +1,11 @@
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: "1Gi"
```
I have a very little understanding about what the DB is actually used for - do you know if 1Gi is a reasonable size?

maulikjs replied:
It's used to store tables and metadata, we have been running 1 Gi in prod for some time now and it hasn't complained.

![image](https://user-images.githubusercontent.com/22432723/101179936-e8560a80-3618-11eb-8c12-f623295fe489.png)

tumido replied:
Same answer as at the Thrift server PR. I've used values based on the internal data hub team expertise. :slightly_smiling_face:

Hue stores mainly it's "user" data and preferences, (it's an old school Django app), query history, sample data cache for individual tables, data autocomplete and hive metadata cache, etc... In our experience this value is enough, though I don't really know how much lower we can go.

vpavlin commented on hue/README.md:
```diff
@@ -0,0 +1,47 @@
+# Cloudera Hue
```
It seems like this does not work without thrifserver, right? If so, can you mention it in the readme and maybe even as a comment in KFDef?

tumido replied:
It can work without thriftserver, but you'd need to connect it to something else (yarn, Hive directly, etc) via modifying the `hue.ini`. Or, you can use it as an S3 explorer. Which is a standalone functionality. In our scenario and set up it requires Thrift server to be available. I'll explain it in the readme. :+1:

tumido replied:
I've added a paragraph in there. WDYT? :slightly_smiling_face:

LaVLaS replied:
This clarification looks good to me.

vpavlin commented on hue/hue/base/params.yaml:
```diff
@@ -0,0 +1,12 @@
+    kind: Secret
+  - path: stringData/DATA_SOURCE_NAME
+    kind: Secret
+  - path: metadata/annotations/volume.beta.kubernetes.io/storage-class
```
This does not template the variable correctly..see https://github.com/opendatahub-io/odh-manifests/blob/master/jupyterhub/jupyterhub/base/params.yaml#L7

vpavlin reviewed (changes requested):
I looks good if the comments I added get fixed!

Thanks!

LaVLaS commented on hue/hue/base/kustomization.yaml:
```diff
@@ -0,0 +1,108 @@
+
+images:
+  - name: hue
+    newName: quay.io/tcoufal/hue
```
@tumido I pushed your personal image to quay.io/opendatahub/hue, can you update the PR to reflect that image

tumido replied:
@LaVLaS sure thing! :+1:  I'll reroute this and relieve my quay account of the Hue duty. :slightly_smiling_face:

LaVLaS commented on hue/hue/base/kustomization.yaml:
```diff
@@ -0,0 +1,108 @@
+kind: Kustomization
+resources:
+  - hive-site-xml-secret.yaml
+  - hue-db-init-job.yaml
```
This file is missing from the commit

tumido replied:
I've removed it and forgot it in the kustomization, thank you! :smile: 

https://github.com/opendatahub-io/odh-manifests/pull/227#discussion_r536092305

LaVLaS commented on hue/README.md:
```diff
@@ -0,0 +1,49 @@
+
+Set session store secret key for Hue web server.
+
+#### s3_endpoint_url
```
Can you add `s3_is_secure` to the parameters in the readme? It would be required for testing this with the unsecure ceph-nano-0 endpoint

tumido replied:
Oh.. I've removed that variable and re-add it. And I forgot to re-add the readme section. Just a sec. Appologies.

tumido replied:
@LaVLaS done

LaVLaS commented:
I only tested the S3 browser and it worked perfectly to view & upload files.

crobby reviewed (approved):
This is working well for me.  I'm not a hue expert, so I'm sure I haven't exercised it much, but I approve.

LaVLaS reviewed (approved):
/lgtm

I tested the s3 browser with no issues read/writing files to ceph-nano.
I tested Hive SQL with the thriftserver component in added in #228 using the opendatahub.io data exploration [tutorial](https://opendatahub.io/docs/advanced-tutorials/data-exploration.html) without any major issues
